From f2a12723d60d51cab09bc64468f33c69f54f8eab Mon Sep 17 00:00:00 2001
From: Carlos Garnacho <carlosg@gnome.org>
Date: Fri, 12 Jul 2013 14:54:04 +0200
Subject: [PATCH 1/2] iconfactory: Check 2x variants more closely before
 attaching

Ensure the 2x variant is actually bigger (fallbacks ensure you
get either 2x or 1x icons) before attaching it to the returned
pixbuf.
---
 gtk/gtkiconfactory.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/gtk/gtkiconfactory.c b/gtk/gtkiconfactory.c
index b97b38d..291c05e 100644
--- a/gtk/gtkiconfactory.c
+++ b/gtk/gtkiconfactory.c
@@ -1740,17 +1740,20 @@ gtk_icon_set_render_icon (GtkIconSet        *icon_set,
   pixbuf = gtk_icon_set_render_icon_internal (icon_set, style, direction,
                                               state, size, widget, detail,
                                               &scale);
-  if (scale == 1)
+  if (pixbuf && scale == 1)
     {
       scale = 2;
       variant = gtk_icon_set_render_icon_internal (icon_set, style, direction,
                                                    state, size, widget, detail,
                                                    &scale);
-      if (variant)
+      if (variant &&
+	  gdk_pixbuf_get_width (variant) > gdk_pixbuf_get_width (pixbuf))
         g_object_set_data_full (G_OBJECT (pixbuf),
                                 "gdk-pixbuf-2x-variant",
                                 variant,
                                 (GDestroyNotify) g_object_unref);
+      else if (variant)
+        g_object_unref (variant);
     }
 
   return pixbuf;
-- 
1.8.3.2

